src/peer_connection.cpp

```diff
@@ -5683,94 +5692,27 @@ namespace libtorrent
 			// from being at or exceeding the limit down to below the limit
 			return;
 		}
-		error_code ec;
-
-		try_read(read_async, ec);
-	}
-
-	size_t peer_connection::try_read(sync_t s, error_code& ec)
-	{
-		TORRENT_ASSERT(is_single_thread());
 		TORRENT_ASSERT(m_connected);
+		if (m_quota[download_channel] == 0) return;
 
-		if (m_quota[download_channel] == 0)
-		{
-			ec = boost::asio::error::would_block;
-			return 0;
-		}
-
-		if (!can_read())
-		{
-			ec = boost::asio::error::would_block;
-			return 0;
-		}
+		int const quota_left = m_quota[download_channel];
+		int const max_receive = (std::min)(buffer_size, quota_left);
 
-		int max_receive = m_recv_buffer.max_receive();
-
-		// only apply the contiguous receive buffer when we don't have any
-		// outstanding requests. When we're likely to receive pieces, we'll
-		// save more time from avoiding copying data from the socket
-		if (m_download_queue.empty())
```

Comment:
the whole shenanigan with null_buffers and looping to drain the socket and try_read() and read_data_impl() was removed and replaced by a single async callback issued from setup_read(). The callback drains the socket (only if the receive buffer was saturated) by first asking for the number of bytes available to be read, and then issuing a single synchronous read() call.